cache/src/main/java/io/envoyproxy/controlplane/cache/SimpleCache.java

```diff
@@ -95,6 +98,25 @@ public Watch createWatch(boolean ads, DiscoveryRequest request) {
 
       Watch watch = new Watch(ads, request);
 
+      HashSet<String> requestedResources = new HashSet<>(request.getResourceNamesList());
```

Comment:
This may not be worth the extra nesting, but you could check that `snapshot != null` before instantiating this set, i.e.:

```java
if (snapshot != null) {
    HashSet<String> requestedResources = new HashSet<>(request.getResourceNamesList());

    if (!knownResourceNames.equals(requestedResources)) {
        // ...
    }
}
```